### PR TITLE
Rewrite product.h

### PR DESCRIPTION
**Pull request by beru**
Hi, this PR rewrites `product.h`. In this way, I think it is somehow simplified.

### CODE REVIEW

**Comment by beru** [comment]
I need to update the code to support old gcc's default ABI ...
https://stackoverflow.com/a/27297025

**Comment by edgarriba** [comment]
What's the status of this?

File: CMakeLists.txt [hunk 1]
```diff
@@ -234,6 +234,10 @@ if(CMAKE_COMPILER_IS_GNUCXX OR MINGW OR
         set(EXTRA_C_FLAGS "${EXTRA_C_FLAGS} -mavx2 -mfma -march=core-avx2")
     endif(USE_AVX2 AND COMPILER_HAS_AVX2_FLAG)
 
+    if(GCC_VERSION_MAJIOR LESS 5)
```
Comment (kecsap) [inline CMakeLists.txt]: I think this is wrong: GCC_VERSION_MAJIOR -> GCC_VERSION_MAJOR

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -15,6 +14,10 @@ add_subdirectory(${googletest_SOURCE_DIR} ${googletest_BINARY_DIR})
 
 enable_testing()
 
+if(GCC_VERSION_MAJIOR LESS 5)
```
Comment (kecsap) [inline test/CMakeLists.txt]: Again: GCC_VERSION_MAJIOR -> GCC_VERSION_MAJOR

File: CMakeLists.txt (hunk 1 above)
Comment (beru) [inline CMakeLists.txt]: Thank you for finding the mistake.

File: test/CMakeLists.txt (hunk 2 above)
Comment (beru) [inline test/CMakeLists.txt]: Thank you for finding the mistake.

File: tiny_dnn/util/macro.h [hunk 3]
```diff
@@ -12,3 +12,10 @@
 #if defined _WIN32 && !defined(__MINGW32__)
 #define CNN_WINDOWS
 #endif
+
+#ifdef _MSC_VER
+#define CNN_MUSTINLINE __forceinline
+#else
+#define CNN_MUSTINLINE __attribute__((always_inline)) inline
```
Comment (beru) [inline tiny_dnn/util/macro.h]: I asked @karandesai-96 to add updated version of macros at #687.
I will rebase this branch once it'll be merged to master.

**Comment by beru** [comment]
> What's the status of this?

@edgarriba It's in abandoned state.

GCC4 default ABI doesn't allow function overloading with `__m128` and `__m256` parameter types. So I guess I need to wrap the types with structs or classes just like original code does...

**Comment by bhack** [comment]
Can we use -fabi-version=4 compiler option ?

**Comment by bhack** [comment]
Or the is the second solution here adopted by Eigen https://stackoverflow.com/questions/5926236/overloading-conflict-with-vector-types-m128-m256-in-gcc

**Comment by bhack** [comment]
Line 31 of https://eigen.tuxfamily.org/dox/SSE_2PacketMath_8h_source.html

**Comment by beru** [comment]
@bhack I think changing abi version is like giving users (who use old gcc) an additional requirement.
So I need to choose the workaround which is not very pleasant from developer's point of view.

**Comment by edgarriba** [comment]
@beru should we close this or are you still working on that?

**Comment by beru** [comment]
@edgarriba I shall close this.

**Pull request closed without merging**